src/generic_event_type.cpp

```diff
@@ -0,0 +1,374 @@
+        generic_event_types::queue_generic_event( pair.first, pair.second );
+    }
+    if( damage != 0 ) {
+        if( target_part.is_valid() ) {
+            target.deal_damage( nullptr, target_part, damage_instance( DT_BASH,
+                                damage ) );
+        } else {
+            for( const bodypart_id &bp : target.get_all_body_parts() ) {
+                target.deal_damage( nullptr, bp, damage_instance( DT_BASH, damage ) );
+            }
+        }
+    }
+    if( weather_change != WEATHER_NULL ) {
+        get_weather().weather_override = weather_change;
+    }
+    if( update_weather ) {
+        get_weather().next_weather = true;
+    }
+
+    target.mod_healthy( healthy );
+    target.mod_rad( radiation );
+    wet_character( target, wet );
+    target.mod_pain( pain );
+    sound( sound_message, sound_effect );
+    target.add_msg_if_player( message );
+}
+
+void generic_event_types::load_pair( const JsonObject &jo, const std::string & )
+{
+    g->generic_events_vector.emplace_back( generic_requirement_type_id( jo.get_string( "require" ) ),
+                                           generic_event_type_id( jo.get_string( "event" ) ) );
+}
+
+void generic_event_types::process_generic_pairs()
+{
+    std::vector<std::pair<time_point, generic_event_type_id>>::iterator queued_event =
+                g->queued_generic_events.begin();
+
+    while( queued_event != g->queued_generic_events.end() ) {
+        if( queued_event->first <= calendar::turn ) {
+            queued_event->second->do_event( );
+            queued_event = g->queued_generic_events.erase( queued_event );
+        } else {
+            ++queued_event;
+        }
+    }
+    for( std::pair<generic_requirement_type_id, generic_event_type_id> require_event :
+         g->generic_events_vector ) {
+        if( require_event.first->test( get_player_character().pos(), get_player_character(),
+                                       WEATHER_NULL ) ) {
+            require_event.second->do_event( );
+        }
+    }
+}
+
+void generic_event_types::reset()
+{
+    generic_event_type_factory.reset();
+}
+
+void generic_event_types::finalize_all()
+{
+    generic_event_type_factory.finalize();
+    for( const generic_event_type &get : generic_event_type_factory.get_all() ) {
+        const_cast<generic_event_type &>( get ).finalize();
+    }
+}
+
+const std::vector<generic_event_type> &generic_event_types::get_all()
+{
+    return generic_event_type_factory.get_all();
+}
+
+void generic_event_types::check_consistency()
+{
+    generic_event_type_factory.check();
+}
+
+void generic_event_types::load( const JsonObject &jo, const std::string &src )
+{
+    generic_event_type_factory.load( jo, src );
+}
+
+void generic_event_types::queue_generic_event( time_duration duration, generic_event_type_id id )
+{
+    g->queued_generic_events.emplace_back( calendar::turn + duration, id );
```

Comment:
Queued events should probably be stored in a heap, not just `push_back`ed onto a vector, for O(log(n)) insertion and handling.  You can use [`std::priority_queue`](https://en.cppreference.com/w/cpp/container/priority_queue).